test-harness/src/tracing_json.rs

```diff
@@ -0,0 +1,119 @@
+use std::collections::HashMap;
+use std::fmt::Display;
+
+use miette::Context;
+use miette::IntoDiagnostic;
+use serde::Deserialize;
+use tracing::Level;
+
+/// A [`tracing`] log event, deserialized from JSON log output.
+#[derive(Deserialize, Debug)]
+#[serde(try_from = "JsonEvent")]
+pub struct Event {
+    /// The event timestamp.
+    pub timestamp: String,
+    /// The level the event was logged at.
+    pub level: Level,
+    /// The log message. May be a span lifecycle event like `new` or `close`.
+    pub message: String,
+    /// The event fields; extra data attached to this event.
+    pub fields: HashMap<String, serde_json::Value>,
+    /// The target, usually the module where the event was logged from.
+    pub target: String,
+    /// The span the event was logged in, if any.
+    pub span: Option<Span>,
+    /// Spans the event is nested in, beyond the first `span`.
+    pub spans: Vec<Span>,
+}
+
+impl Event {
+    /// Get an iterator over this event's spans, from the inside out.
+    pub fn spans(&self) -> impl Iterator<Item = &Span> {
+        self.span.iter().chain(self.spans.iter())
+    }
+}
+
+impl Display for Event {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} {}", self.level, self.target)?;
+        let spans = itertools::join(self.spans(), ">");
+        if !spans.is_empty() {
+            write!(f, " [{spans}]")?;
+        }
+        write!(f, ": {}", self.message)?;
+        if !self.fields.is_empty() {
+            write!(f, " {}", display_map(&self.fields))?;
+        }
+        Ok(())
+    }
+}
+
+impl TryFrom<JsonEvent> for Event {
+    type Error = miette::Report;
+
+    fn try_from(event: JsonEvent) -> Result<Self, Self::Error> {
+        Ok(Self {
+            timestamp: event.timestamp,
+            level: event
+                .level
+                .parse()
+                .into_diagnostic()
+                .wrap_err_with(|| format!("Failed to parse tracing level: {}", event.level))?,
+            message: event.fields.message,
+            fields: event.fields.rest,
+            target: event.target,
+            span: event.span,
+            spans: event.spans,
+        })
+    }
+}
+
+#[derive(Deserialize)]
+struct JsonEvent {
+    timestamp: String,
+    level: String,
```

Comment:
i think you can derive `Deserialize` for a field with an enum type (e.g. `Level`)

Like, you can still keep two types (e.g. `Event` and `JSONEvent`) if you want to keep your data model separate from your JSON interface.  The main reason I'm suggesting pushing the `Level` parsing into the JSON deserializer is so that parsing only happens in one place (JSON deserialization) instead of two places (JSON deserialization and `try_from`).

Reply:
Unfortunately [`Level`](https://docs.rs/tracing/latest/tracing/struct.Level.html) is a third-party type which doesn't implement `Deserialize`, and Rust doesn't allow orphan implementations.

Although `try_from` augments the deserialization, it is still called in the `Deserialize` trait — that is, we can do `serde_json::from_str<Event>(...)` and it will "just work".

I could create a wrapper `struct LevelDeserialize(Level);` and implement `Deserialize` for that, but [implementing `Deserialize`](https://serde.rs/impl-deserialize.html) is an enormous pain involving implementing a [`Visitor`](https://docs.rs/serde/1.0.188/serde/de/trait.Visitor.html) with 28 methods (one for each type in the [`serde` data model](https://serde.rs/data-model.html)). Overall, `TryFrom` is the easiest way I've found to do this sort of thing.